srcs/app/server/Server.cpp

```diff
@@ -13,40 +14,89 @@ Server::~Server() {
 	close(listen_sd_);
 }
 
-// Delegate how to handle connection responsability to Server
-void	Server::AddConnection(int sd) {
+void	Server::BindListeningSocket_() {
+	if (listen_sd_ < 0) {
+		throw std::runtime_error(std::strerror(errno));
+	}
+	if (fcntl(listen_sd_, F_SETFL, O_NONBLOCK) < 0) {
+		throw std::runtime_error(std::strerror(errno));
+	}
+
+	struct sockaddr_in addr;
+	addr.sin_family = AF_INET;  // IPv4
+	addr.sin_port = htons(settings_.listen_port);
+	addr.sin_addr.s_addr = htonl(settings_.listen_address);
+	std::memset(addr.sin_zero, 0, sizeof(addr.sin_zero));
+
+	int on = 1;
+	if (setsockopt(listen_sd_, SOL_SOCKET, SO_REUSEADDR, &on, sizeof(on)) < 0) {
+		throw std::runtime_error(std::strerror(errno));
+	}
+
+	if (bind(listen_sd_, (struct sockaddr *)&addr, sizeof(addr)) < 0) {
+		throw std::runtime_error(std::strerror(errno));
+	}
+
+	if (listen(listen_sd_, SOMAXCONN) < 0) {
+		throw std::runtime_error(std::strerror(errno));
+	}
+}
+
+void	Server::AddConnection_(int sd) {
 	HttpRequestHandler *handler = new HttpRequestHandler(settings_);
 	HttpRequest *request = new HttpRequest();
 	Connection *connection = new Connection(sd, handler, request);
 	connections_.insert(std::make_pair(sd, connection));
+	fdSets_->addToReadSet(sd);
 }
 
-void	Server::RemoveConnection(int sd) {
+void	Server::RemoveConnection_(int sd) {
+	fdSets_->removeFd(sd);
 	close(sd);
 	delete connections_[sd];
 	connections_.erase(sd);
 }
 
-int		Server::GetListeningSocket() const {
-	return listen_sd_;
+void	Server::AcceptNewConnection() {
+	int new_sd = accept(listen_sd_, NULL, NULL);
+	if (new_sd < 0) {
+		throw std::runtime_error(std::strerror(errno));
+	}
+	if (fcntl(new_sd, F_SETFL, O_NONBLOCK) < 0) {
+		throw std::runtime_error(std::strerror(errno));
+	}
+	AddConnection_(new_sd);
 }
 
 bool	Server::HasConnection(int sd) {
 	return connections_.count(sd) > 0;
 }
 
-ReceiveRequestStatus::Type	Server::ReceiveRequest(int sd) {
+void	Server::ReceiveRequest(int sd) {
 	std::map<int, Connection *>::iterator it = connections_.find(sd);
 	if (it == connections_.end()) {
-		return ReceiveRequestStatus::kFail;
+		RemoveConnection_(sd);
+	}
```

Comment:
If we haven't found sd in our connections we remove the connection and **fall through**. Then we dereference it at the end... isn't that UB (undefined behaviour)?

Reply:
Good catch, @Damrod.

In theory, this line will never be executed (because we call `WebServer::FindConnectionServer_()` before calling this method) and was added as a precaution.

We could throw a std::logic_error or std::invalid_argument and fail fast if we update this part in the future.

Reply:
Good catch on my side too! this will never be executed as @gbudau says but, as discussed with him, this guard could make sens if you unit test this class. I will keep it, if you don't mind, since it does not bother much having it